R/caretStack.R

```diff
@@ -55,80 +54,95 @@ caretStack <- function(all.models, ...){
 #' \dontrun{
 #' library("rpart")
 #' models <- caretList(
-#'   x=iris[1:100,1:2],
-#'   y=iris[1:100,3],
-#'   trControl=trainControl(method="cv"),
-#'   methodList=c("rpart", "glm")
+#'   x = iris[1:100, 1:2],
+#'   y = iris[1:100, 3],
+#'   trControl = trainControl(method = "cv"),
+#'   methodList = c("rpart", "glm")
 #' )
-#' meta_model <- caretStack(models, method="lm")
-#' RMSE(predict(meta_model, iris[101:150,1:2]), iris[101:150,3])
+#' meta_model <- caretStack(models, method = "lm")
+#' RMSE(predict(meta_model, iris[101:150, 1:2]), iris[101:150, 3])
 #' }
 predict.caretStack <- function(
-  object, newdata=NULL,
-  se=FALSE, level=0.95,
-  return_weights=FALSE,
-  na.action=na.omit,
-  ...){
+    object, newdata = NULL,
+    se = FALSE, level = 0.95,
+    return_weights = FALSE,
+    na.action = na.omit,
+    ...) {
   stopifnot(is(object$models, "caretList"))
   type <- extractModelTypes(object$models)
 
-  #preds <- do.call(predict, c(list(object=object$models, newdata=newdata), modelList_predict_params)) TODO keep this around for future use
-  preds <- predict(object$models, newdata=newdata, na.action=na.action)
-  if(type == "Classification"){
-    out <- predict(object$ens_model, newdata=preds, na.action=na.action, ...)
-    # Need a check here
-    if(class(out) %in% c("data.frame", "matrix")){
-      # Return probability predictions for only one of the classes
-      # as determined by configured default response class level
-      est <- out[, getBinaryTargetLevel(), drop = TRUE]
-    } else{
-      est <- out
+  # preds <- do.call(predict, c(list(object=object$models, newdata=newdata), modelList_predict_params)) TODO keep this around for future use
+  preds <- predict(object$models, newdata = newdata, na.action = na.action)
+
+  if (type == "Classification") {
+    column_names <- colnames(preds)
+    num_classes <- length(levels(object$models[[1]]$pred$obs))
+    if (getMulticlassExcludedLevel() >= 1 && getMulticlassExcludedLevel() <= num_classes) {
+      classes_included <- levels(object$models[[1]]$pred$obs)[-getMulticlassExcludedLevel()]
+    } else {
+      classes_included <- levels(object$models[[1]]$pred$obs)
     }
-  } else{
-    est <- predict(object$ens_model, newdata=preds, ...)
+    pattern <- paste(classes_included, collapse = "|")
+    # Remove columns that are associated with the class that was excluded
+    filtered_column_names <- grep(pattern, column_names, value = TRUE)
```

Comment:
`grep(pattern, column_names, value = TRUE)` can you spend some time thinking about cased where this could go wrong?  E.g. could you have a `classes_included`  with the character `|` in it?  (I think you could, if you had a target for a multiclass problem where one of the levels has that character).

Reply:
Yes, you are right. However, I tried to reproduce this case and the `train` function from the `caret` package produced an exception because the character "|" is not a valid character for a variable name in R:

"Error: At least one of the class levels is not a valid R variable name"

So it will not be possible to include a class level that contains "|" in its name. Therefore, I think we don't need to change this. What do you think?

Reply:
awesome, that makes sense.

because I'm a little paranoid, can you add a unit tests that checks this case, and expects an error if you try to train a caret model with a variable that has `|` in the name?  Just so we don't get caught in the future if this behavior changes.

Reply:
It's a good idea! I will add a test to check this.